### PR TITLE
[ONNX]Fix  export of flatten

### PR DESCRIPTION
**Pull request by KsenijaS**
Shape is passed to _reshape_to_tensor as a Constant and cannot infer shape of the input when model is exported with dynamic axes set. Instead of a Constant pass output of a subgraph Shape-Slice-Concat to compute the shape for the Reshape node in _reshape_to_tensor function.

### CODE REVIEW

File: torch/onnx/symbolic_opset11.py [hunk 1]
```diff
@@ -663,14 +663,16 @@ def flatten(g, input, start_dim, end_dim):
...
         end_dim = dim + end_dim
-    input_dims = input.type().sizes()
-    output_dims = []
-    for i in range(0, dim):
-        if start_dim < i and end_dim >= i:
-            output_dims[start_dim] = output_dims[start_dim] * input_dims[i]
-        else:
-            output_dims.append(input_dims[i])
-    shape = g.op("Constant", value_t=torch.LongTensor(output_dims))
+    input_size = g.op("Shape", input)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py:666]: What's the difference between opset9 and opset11? Could we abstract out the common part, as it seems there are a lot of repeated code.

File: test/onnx/test_utility_funs.py [hunk 2]
```diff
@@ -675,6 +675,29 @@ def forward(self, x):
...
+
+        x_ort = torch.randn(5, 5, 4, 5)
+        ort_sess = onnxruntime.InferenceSession(f.getvalue())
+        ort_inputs = {ort_sess.get_inputs()[0].name : x_ort.cpu().numpy()}
+        ort_outs = ort_sess.run(None, ort_inputs)[0]
+
+        y = model(x_ort)
+        pytorch_outs = y.cpu().numpy()
+
+        [np.testing.assert_allclose(out, ort_out, rtol=1e-3, atol=1e-7) for out, ort_out in zip(pytorch_outs, ort_outs)]
```
Comment (BowenBao) [inline test/onnx/test_utility_funs.py:699]: Should we put this test in `test_pytorch_onnx_onnxruntime.py`? It looks it's just doing the usual stuff: export and run model in ort and compare results with pytorch.

File: torch/onnx/symbolic_opset9.py [hunk 3]
```diff
@@ -1982,15 +1982,18 @@ def flatten(g, input, start_dim, end_dim):
...
-            output_dims[start_dim] = output_dims[start_dim] * input_dims[i]
-        else:
-            output_dims.append(input_dims[i])
-    shape = g.op("Constant", value_t=torch.LongTensor(output_dims))
-    p = _reshape_from_tensor(g, input, shape)
+    input_size = g.op("Shape", input)
+    slice1 = sym_help._slice_helper(g, input_size, axes=[0], starts=[0], ends=[start_dim])
+    slice2 = sym_help._slice_helper(g, input_size, axes=[0],  starts=[start_dim], ends=[end_dim+1])
+    slice2 = g.op("ReduceProd", slice2)
+    slices = [slice1, slice2]
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:1989]: I'm wondering if we can use `-1` for one of the slices, like `[slice1, torch.tensor([-1])]` and `[slice1, slice2, torch.tensor([-1])]`.

File: test/onnx/test_utility_funs.py (hunk 2 above)
Comment (KsenijaS) [reply test/onnx/test_utility_funs.py:699]: the reason why I have put the test in test_utility_funs is to check if model exported with input of shape (3, 5, 4, 5) could run in ort with input of shape (5, 5, 4, 5). I don't think I can do that in test_pytorch_onnx_onnxruntime.py.

File: torch/onnx/symbolic_opset11.py (hunk 1 above)
Comment (KsenijaS) [reply torch/onnx/symbolic_opset11.py:666]: I think the code is the same for opset 9 and 11, don't know why. ONNX Flatten between opset 9 and 11 differs only that opset 11 can have negative axes

File: test/onnx/test_utility_funs.py (hunk 2 above)
Comment (BowenBao) [reply test/onnx/test_utility_funs.py:701]: actually you can :)
please refer to this example in `test_pytorch_onnx_onnxruntime.py` https://github.com/pytorch/pytorch/blob/6debc28964872e26724764e213c40db366327f95/test/onnx/test_pytorch_onnx_onnxruntime.py#L445

File: torch/onnx/symbolic_opset9.py (hunk 3 above)
Comment (KsenijaS) [reply torch/onnx/symbolic_opset9.py]: I don't think it would work. I'm getting this error: onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Node (Concat_98) Op (Concat) [ShapeInferenceError] All inputs to Concat must have same rank

File: torch/onnx/symbolic_helper.py [hunk 4]
```diff
@@ -437,6 +437,27 @@ def assert_training_mode(op_mode, op_name):
                       op_mode + " mode. The model will be exported in " +
                       training_mode + ", as specified by the export mode.")
 
+def _flatten_helper(g, input, start_dim, end_dim, dim):
+    # use Reshape for cases where the output shape is not 2D
+    if not input.isCompleteTensor():
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:442]: no longer needed, since we are dynamically operating on shape produced by "Shape" node.

File: torch/onnx/symbolic_opset9.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py]: yea i forgot to put the tensor value in Constant node. How about 
`[slice1, g.op("Constant", value_t=torch.tensor([-1], dtype=torch.long))]` 
and 
`[slice1, g.op("Constant", value_t=torch.tensor([-1], dtype=torch.long)), slice3]`?

File: torch/onnx/symbolic_helper.py [hunk 5]
```diff
@@ -437,6 +437,27 @@ def assert_training_mode(op_mode, op_name):
...
+        return _unimplemented("flatten",
+                              "input size not accessible "
+                              "(consider using reshape op instead of flatten op to export to ONNX)")
+
+    input_size = g.op("Shape", input)
+    slice1 = _slice_helper(g, input_size, axes=[0], starts=[0], ends=[start_dim])
+    slice2 = _slice_helper(g, input_size, axes=[0], starts=[start_dim], ends=[end_dim + 1])
+    slice2 = g.op("ReduceProd", slice2)
+    slices = [slice1, slice2]
+    if end_dim < dim - 1:
```
[the remark this answers is not in the document]
Comment (KsenijaS) [reply torch/onnx/symbolic_helper.py:452]: yes end_dim can be -1, that's why only if end_dim < dim-1 slice 3 is created

[the remark this answers is not in the document]
Comment (yaeldekel) [reply torch/onnx/symbolic_helper.py:452]: but when `end_dim` is -1 `end_dim < dim-1` is true. I guess `slice3` here will still do the right thing, since it will have `starts=0` and `ends=dim`, but if it is not needed then I think we should add `end_dim != -1` to the condition in the "if".

[the remark this answers is not in the document]
Comment (KsenijaS) [reply torch/onnx/symbolic_helper.py:452]: that's why there is if end_dim < 0 :end_dim = dim + end_dim in flatten in symbolic_opset 9 and 11:
for example: https://github.com/pytorch/pytorch/pull/40418/files#diff-e5e8501daccb885bc428bcf6c54fd98eL664

File: torch/onnx/symbolic_opset9.py [hunk 6]
```diff
@@ -1977,22 +1977,8 @@ def flatten(g, input, start_dim, end_dim):
         return g.op("Flatten", input, axis_i=start_dim)
```
[the remark this answers is not in the document]
Comment (KsenijaS) [reply torch/onnx/symbolic_opset9.py:1977]: yes it can. Good catch

[the remark this answers is not in the document]
Comment (yaeldekel) [reply torch/onnx/symbolic_opset9.py:1982]: Thanks for changing this! 
I think it should be `dim`, not `input_dim`.

File: torch/onnx/symbolic_opset11.py [hunk 7]
```diff
@@ -648,29 +648,20 @@ def im2col(g, input, kernel_size, dilation, padding, stride):
...
 def flatten(g, input, start_dim, end_dim):
     dim = input.type().dim()
+    if dim is None:
+        return _unimplemented("dim",
+                              "ONNX and PyTorch use different strategies to split the input. "
+                              "Input rank must be known at export time.")
+
     # use ONNX's Flatten operator for cases where the output shape is 2D
     if start_dim == 1:
         if (end_dim == -1 or (end_dim is not None and end_dim == dim - 1)):
```
Comment (yaeldekel) [inline torch/onnx/symbolic_opset11.py:658]: For this case, isn't it ok if `dim` is `None`? 
Same question for opset9.

Comment (KsenijaS) [reply torch/onnx/symbolic_opset11.py:658]: I don't think end_dim = None-1 is valid statement

Comment (yaeldekel) [reply torch/onnx/symbolic_opset11.py:658]: I don't think it is valid either. But if `end_dim` is -1 or `None`, then can't it be allowed for `dim` to be `None`?

Comment (KsenijaS) [reply torch/onnx/symbolic_opset11.py:658]: I don't think end_dim can be None, if not specified default value is -1. I don't think this line is correct, maybe the original author meant dim is not None instead end_dim.

Comment (yaeldekel) [reply torch/onnx/symbolic_opset11.py]: I think `dim is not None` makes more sense than `end_dim is not None`. Should we change it?
And if so, the check above that returns `unimplemented(...)` should be done after this if.

Comment (KsenijaS) [reply torch/onnx/symbolic_opset11.py]: I just removed it. The above check for dim should cover it and other cases as well.

File: torch/onnx/symbolic_helper.py [hunk 8]
```diff
@@ -441,6 +441,19 @@ def assert_training_mode(op_mode, op_name):
...
+    input_size = g.op("Shape", input)
+    slice1 = _slice_helper(g, input_size, axes=[0], starts=[0], ends=[start_dim])
+    slices = [slice1, g.op("Constant", value_t=torch.tensor([-1], dtype=torch.long))]
+    if end_dim < dim - 1:
+        slice3 = _slice_helper(g, input_size, axes=[0], starts=[end_dim + 1], ends=[dim])
+        slices = [slice1, g.op("Constant", value_t=torch.tensor([-1], dtype=torch.long)), slice3]
+
+    final_shape = g.op("Concat", *slices, axis_i=0)
+    from torch.onnx.symbolic_opset9 import _reshape_from_tensor
+    p = _reshape_from_tensor(g, input, final_shape)
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:454]: nit: `return _reshape_from_tensor(g, input, final_shape)`?

**Review by BowenBao — COMMENTED** [review]
please check ci failures, are they related to setting slice2 as -1?

**Review by BowenBao — APPROVED** [review]
LGTM

**Comment by KsenijaS** [comment]
@houseroad CI failures are not caused by this PR. Can you please take a look? Thanks

**Comment by KsenijaS** [comment]
@houseroad CI is green! can you please take a look? Thanks!

**Comment by KsenijaS** [comment]
@houseroad Can you please take a look?

**Review by houseroad — APPROVED** [review]
